library/std/src/canon.qs

```diff
@@ -246,6 +247,51 @@ namespace Microsoft.Quantum.Canon {
         return snd;
     }
 
+    /// # Summary
+    /// Exchanges the labels of two qubits. This can function like a SWAP gate, but without needing to
+    /// perform any operations on the qubits themselves. The relabeling is done purely in qubit ID management.
+    ///
+    /// # Input
+    /// ## q0
+    /// The first qubit to swap.
+    /// ## q1
+    /// The second qubit to swap.
+    ///
+    /// # Remarks
+    /// This operation is useful when you need to swap qubits in a way that does not incur any quantum operations.
+    /// Note that when compiling for execution on hardware with limited qubit connectivity, this operation
+    /// may not result in any changes to qubit adjacency and a SWAP gate may still be required.
+    operation SwapLabels(q0 : Qubit, q1 : Qubit) : Unit {
+        Relabel([q0, q1], [q1, q0]);
+    }
+
+    /// # Summary
+    /// Relabels the qubits in the `current` array with the qubits in the `updated` array. The `updated` array
+    /// must be a valid permutation of the `current` array.
+    ///
+    /// # Input
+    /// ## current
+    /// Array of qubits to be relabeled.
+    /// ## updated
+    /// Array of qubits with which to relabel the `current` array.
+    ///
+    /// # Remarks
+    /// This operation is useful when you need to relabel qubits in a way that does not incur any quantum operations.
+    /// Note that when compiling for execution on hardware with limited qubit connectivity, this operation
+    /// may not result in any changes to qubit adjacency and one or more SWAP gates may still be required.
+    ///
+    /// # Example
+    /// The following example demonstrates how to relabel qubits in a register:
+    /// ```qsharp
+    /// use qubits = Qubit[3];
+    /// let newOrder = [qubits[2], qubits[0], qubits[1]];
+    /// Relabel(qubits, newOrder);
+    /// ```
+    /// After this operation, any use of `qubits[0]` will refer to the qubit that was originally `qubits[2]`, and so on.
+    operation Relabel(current : Qubit[], updated : Qubit[]) : Unit {
```

Comment:
What about `SwapLabel` and `SwapLabels` for single vs array relabel? I'm worried about cognitive load of having two very different names depending on the args.

Reply:
The feedback I got from folks (and pattern I saw in some research papers) was to call this permuting of identifiers "relabeling." So for permutations I think `Relabel` is the right term to connect it to the quantum concept. The issue is what to call the single qubit special case, which behaves like a software SWAP instead of a quantum SWAP. One option could be dropping the special case entirely and requiring users to write `Relabel([q0, q1], [q1, q0])` but that doesn't quite feel right to me. I like having a drop in, "free" replacement for `SWAP` I just don't know what to call it to connect to the concept of relabeling given that `Relabel` is taken by the general API.